### PR TITLE
Enable clean / extension-less url

### LINKED ISSUE #413: Allow using extension-less URLs
### Is this a bug report?

No.

### Have you read the [Contributing Guidelines](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md)?

Yes.

### Environment

Any.

### Expected Behavior

Since [Docusaurus has been designed to work well with GitHub Pages](https://docusaurus.io/docs/en/publishing.html#using-github-pages), it would be ideal if the links it generates via markdown references also supported extension-less URLs.

Note GitHub supports this out-of-the box: simply remove the `.html` suffix from any published pages and it just works.

### Actual Behavior

Docusaurus always creates links with the `.html` extension appended, and there is no way to change that.

In practical terms, this means there's no chance of using GitHub Pages' feature of removing file extension suffixes from URLs without sacrificing consistency.

### Reproducible Demo

Example:
* https://prettier.io/docs/en/install.html
* https://prettier.io/docs/en/install

As a side note, also note how `index.html` can also be removed (which in conjunction with #323 would allow for _clean_ URLs):
* https://prettier.io/docs/en/index.html
* https://prettier.io/docs/en/

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

Closes #413

This is a feature request in issue #413, and https://github.com/facebook/Docusaurus/issues/478#issuecomment-369353099

Some of Docosaurus URL look very cryptic/un-clean. Instead of `blog/a.html`, `blog/b.html`, and `blog/c.html`, user should be able to choose if they prefer to see `blog/a`, `blog/b` and `blog/c`

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Test Plan
- [X]  New config `cleanUrl` in `siteConfig.js`, 
If `true`, all linking from each page will be without `.html` except custom pages that has hardcoded linking value.

Before:
before-link

After:
after-link

- [x]  `yarn start` will allow extension-less url (server routing)
Which means request to `docs/a.html` is the same as `docs/a`
**Test `yarn start` server routing on development server**
```
cd website
yarn start
```

`http://localhost:3000/docs/en/installation`
dev-1

`http://localhost:3000/docs/en/installation.html`
dev-2

- [X]  `yarn build` will generate extra file when `cleanUrl` at `siteConfig.js` is true

**Test `yarn build` to see generated files**
```
cd website
yarn build
tree -a -I "css|CNAME|.circleci|img|js" build/Docusaurus/ > build.txt
```

Results are expected
Full generated files structure can be seen at https://gist.github.com/endiliey/cd184c237bc1c0b544677ad19cc8ece7
```
build/Docusaurus/
├── about-slash
│   └── index.html
├── about-slash.html
├── blog
│   ├── 2017
│   │   └── 12
│   │       └── 14
│   │           ├── introducing-docusaurus
│   │           │   └── index.html
│   │           └── introducing-docusaurus.html

├── docs
│   ├── 1.0.11
│   │   ├── api-pages
│   │   │   └── index.html
│   │   ├── api-pages.html
```

- [x] Test production build (static files) in local webserver using https://www.npmjs.com/package/serve
```
cd website
yarn build
serve build/Docosaurus
```

serve

`http://localhost:5000/docs/en/site-preparation`
prod-1-doc

`http://localhost:5000/docs/en/site-preparation.html`
prod-with-html

`http://localhost:5000/blog/2018/04/30/How-I-Converted-Profilo-To-Docusaurus`
prod-blog-clean

`http://localhost:5000/blog/2018/04/30/How-I-Converted-Profilo-To-Docusaurus.html`
prod-blog-html

`http://localhost:5000/en/about-slash`
prod-custom-clean

`http://localhost:5000/en/about-slash.html`
prod-custom-html

All seems to be working fine. Every url is accessible with and without .html

**Alternatively, you can test production build in Netlify**
https://deploy-preview-677--docusaurus-preview.netlify.com/

### CODE REVIEW

**Comment by ericnakagawa** [comment]
This is a neat feature. Thanks for supporting the old urls as well!

**Comment by yansern** [comment]
Just tried this out myself while I was testing the other non-related issue! Great work!

**Comment by endiliey** [comment]
I am not in a rush with this PR, i'm out of reach from my laptop until this weekend because i'm currently hospitalized. But i have tested it on dev & prod build in local web server before and there is no problem yet.

But note that custom pages (e.g: pages/en/index.js)  with hardcoded link still link to docs/installation.html

https://github.com/facebook/Docusaurus/blob/a603aff193096195bd805b5755f662d981d0857e/website/pages/en/index.js#L51

Do we want to change it to extension-less one ?

**Comment by JoelMarcey** [comment]
Hi @endiliey 

> I am not in a rush with this PR, i'm out of reach from my laptop until this weekend because i'm currently hospitalized.

Best wishes to you. 🙏 

> But note that custom pages (e.g: pages/en/index.js) with hardcoded link still link to docs/installation.html

We can change those in a separate PR, I think.

**Comment by endiliey** [comment]
I am ready anytime. I'll leave it to your decision. Waiting for @yangshun approval is also a good idea 😉

**Comment by yangshun** [comment]
I'll give it a thorough look tonight! One concern I have is that will the blog comments will show different contents depending on whether it's the index.html version or the clean URL version? Not sure if FB treats them as canonical URLs.

**Comment by endiliey** [comment]
I just tested that and it is actually the same url. Regardless of `.html` or `/index.html` or `/`. The data-href from facebook comment is not the current url and always point to the clean url (if siteConfig `cleanUrl` is true

I just added 'Wow, awesome' to netlify preview and you can see it at

https://deploy-preview-677--docusaurus-preview.netlify.com/blog/2017/12/14/introducing-docusaurus/index.html

https://deploy-preview-677--docusaurus-preview.netlify.com/blog/2017/12/14/introducing-docusaurus.html

https://deploy-preview-677--docusaurus-preview.netlify.com/blog/2017/12/14/introducing-docusaurus/

File: docs/api-site-config.md [hunk 1]
```diff
@@ -167,6 +167,8 @@ h1 {
 
 `wrapPagesHTML` - Boolean flag to indicate whether `html` files in `/pages` should be wrapped with Docusaurus site styles, header and footer. This feature is experimental and relies on the files being `html` fragments instead of complete pages. It inserts the contents of your `html` file with no extra processing. Defaults to `false`.
 
+`cleanUrl` - If `true`, allow URLs with no `html` extension. Example: request to URL https://docusaurus.io/docs/installation will returns the same result as https://docusaurus.io/docs/installation.html.
```
Comment (JoelMarcey) [inline docs/api-site-config.md]: Can you move this up to be alphabetical with the rest of the optional fields?

Comment (endiliey) [reply docs/api-site-config.md]: I can. But i'm out of reach from my pc until this weekend. I'll do it later if you don't mind. Sorry!

Comment (JoelMarcey) [reply docs/api-site-config.md]: Oh - no worries. I can do it for you 👍

**Comment by endiliey** [comment]
PR updated. Rebased on top of 1.1.4

Not sure if we want to remove the hardcoded linking values in Docusaurus custom pages if this is merged. Example:
https://github.com/facebook/Docusaurus/blob/b7be85843af2a07ead183dd8a467ebcf83e9a3e3/website/pages/en/index.js#L51

https://github.com/facebook/Docusaurus/blob/b7be85843af2a07ead183dd8a467ebcf83e9a3e3/website/pages/en/index.js#L117

**Comment by yangshun** [comment]
@endiliey Could we try to come up with a way to reduce the amount of duplication in the code?

**Comment by endiliey** [comment]
@yangshun refactor done ! Sorry for the wait. I've tested this on Windows as well.

**Comment by JoelMarcey** [comment]
Let's do this!

**Pull request merged**